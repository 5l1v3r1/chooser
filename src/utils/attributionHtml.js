function attributionHtml(form, ccLicenseURL, shortLicenseName) {
    const baseAssetsPath = 'https://search.creativecommons.org/static/img'
    const imgLink = `<a href="${form.workLocation}">"${form.workTitle}"</a>`
    let creator = ''
    if (form.attributeToName && form.attributeToURL) {
        creator = `<span> by <a href="${form.attributeToURL}">${form.attributeToName}</a></span>`
    } else if (form.attributeToName && !form.attributeToURL) {
        creator = `<span> by <span>${form.attributeToName}</span></span>`
    }
    const licenseLink = ` is licensed under <a href="${ccLicenseURL}" style="margin-right: 5px;">${shortLicenseName.toUpperCase()}</a>`

    let licenseIcons = `<img style="height: inherit;margin-right: 3px;display: inline-block;" src="${baseAssetsPath}/cc_icon.svg" />` // eslint-disable-line global-require, import/no-dynamic-require
    if (shortLicenseName) {
        licenseIcons += shortLicenseName.slice(3, shortLicenseName.length - 4).split('-').map(license =>
            `<img style="height: inherit;margin-right: 3px;display: inline-block;" src="${baseAssetsPath}/cc-${license.toLowerCase()}_icon.svg" />` // eslint-disable-line global-require, import/no-dynamic-require
        ).join('')
    }

    const licenseImgLink = `<a href="${ccLicenseURL}" target="_blank" rel="noopener noreferrer" style="display: inline-block;white-space: none;opacity: .7;margin-top: 2px;margin-left: 3px;height: 22px !important;">${licenseIcons}</a>`
    return `<p style="font-size: 0.9rem;font-style: italic;">${imgLink}${creator}${licenseLink}${licenseImgLink}</p>`
}

export default attributionHtml
