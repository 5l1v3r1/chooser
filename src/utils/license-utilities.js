const CC0Attributes = { BY: false, NC: false, ND: false, SA: false }
const CCBYAttributes = { BY: true, NC: false, ND: false, SA: false }
const defaultAttributes = { BY: undefined, NC: undefined, ND: undefined, SA: undefined }
const visibleSetters = {
    FSCC0: ['FS', 'DD', 'CW', 'AD'],
    FSBY: ['FS', 'DD', 'AD'],
    FS: {
        true: ['FS', 'DD', 'AD'],
        false: ['FS', 'BY', 'NC', 'ND', 'SA', 'AD']
    },
    BY: {
        true: ['FS', 'BY', 'NC', 'ND', 'SA', 'AD'],
        false: ['FS', 'BY', 'NC', 'ND', 'SA', 'CW', 'AD']
    },
    ND: {
        true: ['FS', 'BY', 'NC', 'ND', 'SA', 'AD']
    }
}
const disabledSetters = {
    // Steps that should be disabled if other steps are selected/not selected
    BY: {
        false: ['NC', 'ND', 'SA']
    },
    ND: {
        true: ['SA']
    }
}

function shortToAttr(shortLicenseName) {
    const short = shortLicenseName
    if (short.includes('CC0')) {
        return { ...CC0Attributes }
    }
    const nc = short.includes('NC')
    const nd = short.includes('ND')
    const sa = short.includes('SA')
    return { ...CCBYAttributes, NC: nc, ND: nd, SA: sa }
}

function attrToShort(attr) {
    if (attr.BY === undefined) return undefined
    if (!attr.BY) { return 'CC0 1.0' }
    let base = 'CC BY'
    if (attr.NC) { base += '-NC' }
    if (!attr.ND && attr.SA) {
        base += '-SA'
    } else if (attr.ND) {
        base += '-ND'
    }
    base += ' 4.0'
    return base
}

function attrToFull(attr) {
    if (!attr.BY) { return 'CC0 1.0 Universal' }
    let base = 'Attribution'
    if (attr.NC) { base += '-NonCommercial' }
    if (!attr.ND && attr.SA) {
        base += '-ShareAlike'
    } else if (attr.ND) {
        base += '-NoDerivatives'
    }
    base += ' 4.0 International'
    return base
}
// function licenseUrl(shortLicenseName) {
//     Returns url to license from short license name with version number (eg. 'CC BY 4.0')
//     TODO: check how it works: it doesn't use '-' to join elements ?
//     if (shortLicenseName.includes('CC0')) {
//         return 'https://creativecommons.org/publicdomain/zero/1.0/?ref=ccchooser'
//     }
//     const short = shortLicenseName.toLowerCase().slice(3, shortLicenseName.length - 4)
//     return 'https://creativecommons.org/licenses/' + short + '/4.0/?ref=ccchooser'
// }
function licenseUrl(attr) {
    // Returns url to license from short license name with version number (eg. 'CC BY 4.0')
    if (attr.BY === false) {
        return 'https://creativecommons.org/publicdomain/zero/1.0/?ref=ccchooser'
    }
    let short = attrToShort(attr).toLowerCase().slice(3)
    short = short.slice(0, short.length - 4)
    return 'https://creativecommons.org/licenses/' + short + '/4.0/?ref=ccchooser'
}

function licenseSlug(shortLicenseName) {
    // Returns lower case slugified string of license name without the version number
    // 'CC BY 4.0' -> 'cc-by'
    return shortLicenseName
        .toLowerCase()
        .replace(' ', '-')
        .slice(0, shortLicenseName.length - 4)
}

function licenseIconsArr(licenseAttributes) {
    if (!licenseAttributes.BY) {
        return ['zero']
    }
    const iconsArray = []
    for (const key in licenseAttributes) {
        if (licenseAttributes[key]) {
            iconsArray.push(key.toLowerCase())
        }
    }
    return iconsArray
}

function generateHTML(attributionDetails, shortLicenseName) {
    const dataForHtmlGeneration = {
        htmlString: '',
        creator: '',
        workTitle: '',
        licenseIconsLink: ''
    }
    dataForHtmlGeneration.htmlString = '<p xmlns:dct="http://purl.org/dc/terms/"' +
        ' xmlns:cc="http://creativecommons.org/ns#"' +
        ' class="license-text">'
    const iconStyle = 'style="height:22px!important;margin-left: 3px;vertical-align:text-bottom;"'
    const baseAssetsPath = 'https://search.creativecommons.org/static/img'
    let licenseIcons = `<img ${iconStyle} src="${baseAssetsPath}/cc_icon.svg" />`
    if (shortLicenseName.includes('CC0')) {
        shortLicenseName = 'CC CC0 1.0'
    }
    licenseIcons += shortLicenseName.slice(3, shortLicenseName.length - 4).split('-').map(license =>
        `<img  ${iconStyle} src="${baseAssetsPath}/cc-${license.toLowerCase()}_icon.svg" />`
    ).join('')
    dataForHtmlGeneration.licenseIconsLink = `<a href="${licenseUrl(shortToAttr(shortLicenseName))}">${licenseIcons}</a>`

    if (attributionDetails.creatorName) {
        const creatorSpan = `<span rel="cc:attributionName">${attributionDetails.creatorName}</span>`
        if (attributionDetails.creatorProfileUrl) {
            dataForHtmlGeneration.creator = `<a rel="cc:attributionURL" href="${attributionDetails.creatorProfileUrl}">${creatorSpan}</a>`
        } else {
            dataForHtmlGeneration.creator = creatorSpan
        }
    }
    if (attributionDetails.workTitle) {
        const workTitleSpan = `<span rel="dc:title">${attributionDetails.workTitle}</span>`
        if (attributionDetails.workUrl) {
            dataForHtmlGeneration.workTitle = `<a rel="cc:attributionURL" href="${attributionDetails.workUrl}">${workTitleSpan}</a>`
        } else {
            dataForHtmlGeneration.workTitle = workTitleSpan
        }
    }
    return dataForHtmlGeneration
}
export {
    defaultAttributes, CC0Attributes, CCBYAttributes, visibleSetters, disabledSetters,
    shortToAttr, attrToShort, attrToFull, licenseUrl, licenseSlug, licenseIconsArr,
    generateHTML
}
