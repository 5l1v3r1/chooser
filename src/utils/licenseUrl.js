function licenseUrl(licenseName) {
    licenseName = licenseName.toLowerCase().slice(3, licenseName.length - 4)
    return 'https://creativecommons.org/licenses/' + licenseName + '/4.0'
}

function isFullName(licenseName) {
    licenseName.toLowerCase()
    if (licenseName.includes("attribution") || licenseName.includes("noderivatives") ||
        licenseName.includes("sharealike") || licenseName.includes("noncommercial")) {
            console.error("Is Full Name")
        return true
    }
    return false
}

export default licenseUrl
