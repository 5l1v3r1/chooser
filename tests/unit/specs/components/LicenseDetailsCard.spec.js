import { mount, createLocalVue, config } from '@vue/test-utils'
import LicenseDetailsCard from '@/components/LicenseDetailsCard.vue'
import { licenseSlug } from '@/utils/license-utilities'
import VueI18n from 'vue-i18n'
import Vuex from 'vuex'

describe('LicenseDetailsCard.vue', () => {
    let wrapper
    let getters
    let state
    let store
    let i18n

    // Always creates a shallow instance of component
    beforeEach(() => {
            const localVue = createLocalVue()
            localVue.use(VueI18n)
            localVue.use(Vuex)
            state = {
                attributionDetails: {
                    creatorName: 'J Doe',
                    creatorProfileUrl: 'www.author.com',
                    workTitle: 'My work',
                    workUrl: 'www.author.com/pic.jpg'
                }
            }
            getters = {
                shortName: state => {
                    return 'foo'
                },
                fullName: state => {
                    return 'fooBar'
                },
                licenseUrl: state => (mode) => {
                    return 'http://example.com'
                },
                iconsList: state => {
                    return ['by', 'sa']
                }
            }
            store = new Vuex.Store({
                state,
                getters
            })
            const messages = require('@/locales/en.json')
            i18n = new VueI18n({
                locale: 'en',
                fallbackLocale: 'en',
                messages: messages
            })

            config.mocks.i18n = i18n

            config.mocks.i18n.$t = (key) => {
                return i18n.messages[key]
            }
            wrapper = mount(LicenseDetailsCard, {
                localVue,
                store,
                i18n
            })
        })
        // Test for DOM elements which must be present
    it('Has the main div tag', () => {
        expect(wrapper.contains('.selected-license-card')).toBe(true)
    })
    it('Has the h3 tag', () => {
        expect(wrapper.contains('h3')).toBe(true)
    })
    it('Has the h4 tag', () => {
        expect(wrapper.contains('h4')).toBe(true)
    })
    it('Has the a tag', () => {
        expect(wrapper.contains('.license-name')).toBe(true)
    })
    it('Has the p tag', () => {
        expect(wrapper.contains('.chooser-selected-description')).toBe(true)
    })
    it('Has the section tag', () => {
        expect(wrapper.contains('.license-visual-info')).toBe(true)
    })
    it('Has the ul tag', () => {
        expect(wrapper.contains('.license-list')).toBe(true)
    })
    it('Has the li tag', () => {
        expect(wrapper.contains('li')).toBe(true)
    })
    it('Has the span tag', () => {
        expect(wrapper.contains('.readable-string')).toBe(true)
    })

    // Tests for computed props and methods
    it('renders without any errors', () => {
        expect(wrapper.isVueInstance()).toBeTruthy()
    })
    it('Check if the slug function returns the correct text', () => {
        expect(wrapper.vm.slug).toBe(licenseSlug('foo'))
    })
    it('Check if the licenseKey function returns the correct text', () => {
        expect(wrapper.vm.licenseKey).toBe(`license-details-card.full-description.${licenseSlug('foo')}`)
    })
    it('Check if the licenseDescription function returns the correct text', () => {
        expect(wrapper.vm.licenseDescription).toBe(`${licenseSlug('foo')}-description`)
    })

    // Snapshot tests
    it('has the expected UI', () => {
        expect(wrapper).toMatchSnapshot()
    })
})
