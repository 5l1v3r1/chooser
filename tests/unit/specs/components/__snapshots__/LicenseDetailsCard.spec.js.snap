// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicenseDetailsCard.vue has the expected UI 1`] = `
<div class="selected-license-card">
  <h3 class="vocab ha h3a">
    license-details-card.heading
  </h3>
  <h4 class="vocab h4b hb"><a href="http://example.com" class="license-name">
      fooBar (foo)
      <a href="http://example.com" target="_blank" rel="noopener noreferrer" class="photo-license-icons"><img src="../assets/license-icons/cc_icon.svg" class="photo-license-icon"> <img src="[object Object]" class="photo-license-icon"><img src="[object Object]" class="photo-license-icon"></a></a></h4>
  <p class="chooser-selected-description"><b>FO</b>
    license-details-card.full-description.fo
  </p>
  <section class="license-visual-info">
    <ul class="license-list"><span name="highlight"><li class="license-list-item by"><span class="readable-string"><b>BY:</b>
                        license-details-card.item-description.by
                    </span></li>
      <li class="license-list-item sa"><span class="readable-string"><b>SA:</b>
                        license-details-card.item-description.sa
                    </span></li>
      </span></ul>
  </section>
</div>
`;
